### ISSUE TITLE
Unable to use stdin in child process using pty

### ISSUE THREAD

**Issue by davehayes** [jhthorsen/mojo-ioloop-readwritefork#13]
When I attempt to write to a child process with this module, I get a set of errors I'm not understanding.  Here is the test code:
```perl
#!/usr/local/bin/perl
use Mojo::Promise;
use Mojo::IOLoop;
use Mojo::IOLoop::ReadWriteFork;
use POSIX;

my $rwf = Mojo::IOLoop::ReadWriteFork->new;
my $buffer = '';
$rwf->conduit({ raw => 1, type => "pty" });

$rwf->on(error => sub {
  my ($rwf, $error) = @_;
  warn "## ($$) ERROR($error)\n";
});
  
$rwf->on(close => sub {
  my ($rwf, $status, $signal) = @_;
  warn "## ($$) CLOSE($status)($signal): $buffer\n";
  Mojo::IOLoop->stop;
});

$rwf->on(read => sub {
  my ($rwf, $readbuf) = @_;
  warn "## ($$) READ ($readbuf)\n";
  $buffer .= $readbuf;
});

my $prom = Mojo::Promise->new();
$prom->then(sub {
  warn "## ($$) Writing data\n";
  $rwf->write("testing one two three", sub {
    $rwf->close('stdin');
  });
})->catch(sub {
  warn "We should never get this!\n";
});


$rwf->on(fork => sub {
  # This is the parent
  my ($rwf) = @_;
  warn "## ($$) Parent: Child PID = ".$rwf->pid."\n";
  $prom->resolve();
});

$rwf->run( sub {
  #  This is the child 
  my ($self) = @_;
  warn "## ($$) Exec cat\n";
  exec('/bin/cat') || warn "Can't exec /bin/cat: $!\n";
  POSIX::_exit();
});

Mojo::IOLoop->start;

```
The output of this script is:
```
[91680] Child starting (CODE(0x80708bf30) )
## (87835) Parent: Child PID = 91680
## (87835) Writing data
[91680] Starting CODE(0x80708bf30) 
[91680] <<< testing one two three
[91680] Child Bad file descriptor
## (87835) ERROR(Read error: Bad file descriptor)
[91680] Child Bad file descriptor
## (87835) ERROR(Read error: Bad file descriptor)
[91680] Child is dead (0/Interrupted system call) 0/1
EV: error in callback (ignoring): Handle is closed at /usr/local/lib/perl5/site_perl/Mojo/IOLoop/ReadWriteFork.pm line 195.
^C
```

The Ctrl-C is because the process never returns, because `close` is never emitted. 

Is this a bug in your or my code? Thanks in advance.

**Comment by jhthorsen** [comment]
From my experience there’s some applications that does not work very well with pty, for reasons I cannot explain. It’s been a while since I wrote the unit tests, but I think I recall having some issues with “cat”.

If this is real code, then I suggest reconsider, since there’s plenty of ways to do “cat” with plain Perl.

**Comment by davehayes** [comment]
I used cat just as a test. The real code is a wrapper object which is a non-blocking replacement of my old blocking object; both of these want to attach a pty to a command, send input, and receive output. My old code works just fine except that it blocks. The above code is just me duplicating the condition which causes the error.

So some data. :) If I change to this:

```perl
my $prom = Mojo::Promise->new();
$prom->then(sub {
  warn "## ($$) Writing data\n";
  $rwf->write("testing one two three", sub {
    Mojo::IOLoop->timer(1 => sub { $rwf->close('stdin'); });
  });
})->catch(sub {
  warn "We should never get this!\n";
});

$rwf->on(fork => sub {
  # This is the parent
  my ($rwf) = @_;
  warn "## ($$) Parent: Child PID = ".$rwf->pid."\n";
  $prom->resolve();
});
```
then cat runs but I still get the errors:

```
[79273] Child starting (CODE(0x80708bfa8) )
## (77314) Parent: Child PID = 79273
## (77314) Writing data
[79273] <<< testing one two three
[79273] Starting CODE(0x80708bfa8) 
[79273] >>> testing one two three
## (77314) READ (testing one two three)
[79273] >>> ## (79273) Exec cat\n
## (77314) READ (## (79273) Exec cat
)
[79273] >>> testing one two three
## (77314) READ (testing one two three)
[79273] Child Bad file descriptor
## (77314) ERROR(Read error: Bad file descriptor)
[79273] Child Bad file descriptor
## (77314) ERROR(Read error: Bad file descriptor)
[79273] Child is dead (0/Interrupted system call) 0/1
EV: error in callback (ignoring): Handle is closed at /usr/local/lib/perl5/site_perl/Mojo/IOLoop/ReadWriteFork.pm line 195.
```

If I never close stdin, no errors, but the application hangs:

```
[39781] Child starting (CODE(0x80708bf18) )
## (36924) Parent: Child PID = 39781
## (36924) Writing data
[39781] <<< testing one two three
[39781] Starting CODE(0x80708bf18) 
[39781] >>> testing one two three
## (36924) READ (testing one two three)
[39781] >>> ## (39781) Exec cat\n
## (36924) READ (## (39781) Exec cat
)
[39781] >>> testing one two three
## (36924) READ (testing one two three)
...never returns...
```
Note that the close event never fires. I can cause the application to stop hanging by doing something like

```perl
my $prom = Mojo::Promise->new();
$prom->then(sub {
  warn "## ($$) Writing data\n";
  $rwf->write("testing one two three", sub {
    Mojo::IOLoop->timer( 1 => sub { Mojo::IOLoop->stop } );
  });
})->catch(sub {
  warn "We should never get this!\n";
});
```

but I can never really detect when the application is done because the close event never fires in any of the cases I have mentioned. 

I will play with this some more but any insight you might have is warmly appreciated as I'm lost as to why this is happening too.

**Comment by jhthorsen** [comment]
Please reopen with a smaller example if this is not fixed in 1.00 release.
